### ISSUE TITLE
[Issue] ConcurrentModificationException

### ISSUE THREAD

**Issue by Moustik2002** [amadornes/MCMultiPart#47]
Hi Client Crash and get this crash log : http://pastebin.com/C3MKNEDR

**Comment by amadornes** [comment]
Do you know what multiparts could be causing this? It'd help narrow things down a lot

**Comment by Moustik2002** [comment]
only things i've used with your mod is glowstone nook and Covers from simplycover mod to hide Refined Storage Cables.

**Comment by bdew** [comment]
I've seen that too. Not sure what parts were placed at that point and i don't think it matters really.

This code is called from chunk rendering threads, so any modification of the parts map (which is a HashBiMap and is not thread safe) at the same time (e.g. due to a packet) will cause a CME.

You should probably either make a copy before iterating or add synchronization. I don't know which would be worse for performace. God i hate java collections sometimes...

**Comment by amadornes** [comment]
@bdew Thing is... I make a copy of the map every single time I modify it to get around CMEs, so I don't know why this would happen :/
I might've missed something, though

**Comment by bdew** [comment]
From a cursory look i see that the map is cleared in readFromNBT, i'm not sure if that could be called while stuff is rendered but it would cause a CME if it does.

**Comment by raoulvdberge** [comment]
Maybe try a `Maps.synchronizedBiMap()`?

**Comment by bdew** [comment]
Another user has reported a CME, it's in my code but it also happens when iterating over the parts map from a rendering thread, so i suspect it's the same underlying issue >.>

http://pastebin.com/nFgX2NYD

**Comment by amadornes** [comment]
Okay, just realized why this issue still occurs. I haven't pushed a proper 1.10 release yet >_>
It was fixed in the first commit after the previous release. From now on you'll need to use getActualState() and getExtendedState() with the world and the position as arguments (I just added those arguments to getActualState, but they've been in getExtendedState for over 2 months).

I'll look into other big issues that are currently open (if any) and push an update.
